Clean up text for second half of DE instruction notebook

allyhawkins opened the pull request:
⚠️ stacked on #596 

Continuing with #573, this PR cleans up the second half of the DE instruction notebook starting with the `Differential Expression Analysis` header until the end. 

A few points of potential discussion:

- I added some text before we dive into pseudo-bulking that discusses the use of the uncorrected raw counts rather than the integrated results that we had been working with up until this point. Do we think this is the right place to talk about this? Do we want to show anything else related to this since we have that assay stored in our experiment? 

- In the pseudo-bulking section, I printed out the `colData` to discuss the addition of the `ncells` column. I however realized I don't talk about the `NA` now found in the QC columns. As I'm filing this, I feel like I should at least include a sentence on why those are now `NA`. 

- Are we okay with the plots that I chose to include? Do we want to include different types of plots to look at the results from the DE analysis? I think its a nice mix of "here's what you could do" but realistically the options are endless and totally dependent on the biological question... 

Here's the rendered report for reference:
[04-differential_expression.nb.html.zip](https://github.com/AlexsLemonade/training-modules/files/10282174/04-differential_expression.nb.html.zip)

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -336,125 +346,162 @@ counts_mtx
+For our experiment we are comparing gene expression between different RMS subtypes, stored in the `diagnosis_group` column of the `colData` in the pseudo-bulked SCE.
+
+```{r deseq object}
+# extract the metadata for each sample from sce as data.frame
 coldata_df <- as.data.frame(colData(tumor_myoblast_sce))
   
 # set up the deseq object 
 deseq_object <- DESeq2::DESeqDataSetFromMatrix(counts(tumor_myoblast_sce),
+                                               # metadata from coldata 
                                                colData = coldata_df,
+                                               # name of column to use for grouping samples 
                                                design = ~ diagnosis_group)
````
We can actually simplify this, as we already have a summarized experiment (SCE is a SE)!
While this makes for a bit of an indirection, not having to convert the colData is a win in my book.

```suggestion
# set up the deseq object 
deseq_object <- DESeq2::DESeqDataSet(tumor_myoblast_sce,
                                     # formula for grouping samples 
                                     design = ~ diagnosis_group)
```

sjspielman replied:
>  as we already have a summarized experiment (SCE is a SE)!

I'd make sure to emphasize this point too during this step to explain why we can jump right in

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -336,125 +346,162 @@ counts_mtx
+Before we can compare gene expression we will need to normalize the counts data.
+If you look at the pseudo-bulked SCE object you should have only seen one assay, the `counts` assay.
+This is because we took the unnormalized counts from the original SCE to create the new pseudo-bulked counts which then become the input to create the `DESeq2` object.
+We don't want to sum the normalized counts because those have each been normalized within their sample and do not account for differences between samples.
```
I would also say this is because you can't just sum logs... but I think this sentence may not be needed? It may be enough to just say that DESeq2 expects raw counts.

sjspielman replied:
It may especially not be needed if we link the bulk materials somewhere.

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
 
-```{r}
-# save our results 
+We can save the new `data.frame` that we have created with the ensembl identifiers, gene symbols, and the `DESeq2` results as a tab separated (`tsv`) file.
````
```suggestion
We can save the new `data.frame` that we have created with the Ensembl identifiers, gene symbols, and the `DESeq2` results as a tab separated (`tsv`) file.
```

sjspielman replied:
I think I might remove `data.frame` --> "data frame" since this isn't referencing that function, just saying the noun "data frame."
Though there are so many kinds of data frames out there that there is a good argument for not doing that!

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
 ```
 
-Now we can get the updated results and combine with the `rowData` from the SCE object to extract the gene symbol. 
+If you look at our `shrink_results` object, we see that the genes are labeled with the ensembl gene identifiers, as those were the rownames of the pseudo-bulked SCE we used as input to build our `DESeq2` object.
````
```suggestion
If you look at our `shrink_results` object, we see that the genes are labeled with the Ensembl gene identifiers, as those were the row names of the pseudo-bulked SCE we used as input to build our `DESeq2` object.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
-# create a data frame with the results 
-deseq_results <- shrink_results %>%
+```{r add gene symbol}
+# convert shrunken results to a data.frame and add ensembl id as a column
````
I don't know if we want to get into the difference between a tibble and a data.frame. But 
```suggestion
# convert shrunken results to a tibble and add Ensembl id as a column
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
 ```
 
+The last thing that we will do is take a look at how many genes are significant.
+Here we will want to use the adjusted p-value, found in the `padj` column of the results, as this p-value accounts for multiple test correction.
````
It isn't technically a p-value anymore, so let's not call it that exactly...
```suggestion
Here we will want to use the adjusted p-value, found in the `padj` column of the results, as this accounts for multiple test correction.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
+Now that we have identified a set of genes that are DE in the tumor myoblasts between ARMS and ERMS subtypes, lets actually take a look at them and see if we can make some informative plots.
+The first plot we'll make is a volcano plot using the [`EnhancedVolcano` package](https://github.com/kevinblighe/EnhancedVolcano).
+This package automatically color codes the points by cutoffs for both significance and fold change and labels many of the significant genes (subject to spacing).
+`EnhancedVolcano` has many, many options, which is a good thing if you don't like all of it's default settings.
```
Putting this comment here, because I can't figure out where else, but I feel like the "fun arrows" point the wrong way. Nothing to be done about it.

sjspielman replied:
```suggestion
`EnhancedVolcano` has many, many options, which is a good thing if you don't like all of its default settings.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+# create a violin plot 
 scater::plotExpression(tumor_sce,
-                       features = genes_to_plot, #PTPRT
+                       features = genes_to_plot, # provide a vector of genes to plot
```
```suggestion
                       features = genes_to_plot, # a vector of genes to plot
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+
+Go ahead and explore some genes on your own! 
+Feel free to plot any of the genes that are identified as significant, found in the DE results table, or your favorite gene.
+Remember, you need to use the ensembl gene identifier to refer to each gene.
```
```suggestion
Remember, you need to use the Ensembl gene identifier to refer to each gene.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+Because we want to show multiple genes here, we are going to add an additional option to `facet_grid()` to include multiple rows in our plot grid, one for each gene of interest.
 
 ```{r}
 # pick a few genes to look at 
````
```suggestion
# pick a couple genes to look at 
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+  facet_grid(cols = vars(celltype_broad), # each celltype is its own column
+             rows = vars(Feature)) + # each feature is its own row
+  theme(strip.text = element_text(size = 7)) + # change the font size of the facet labels
+  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) # change the size of the legend colors
```
As long as we are getting into it, do we want to update the legend title?
```suggestion
  guides(colour = guide_legend(
      title = "Subtype", # update the legend title
     override.aes = list(size = 3, alpha = 1)) # change the size of the legend colors
 ) 
```

sjspielman replied:
Noting that the long comments here might be too long, so we might move them to above the lines instead of at the end.

jashapiro reviewed (commented):
Despite my many comments here, I think this looks very good.

Most of my comments are pretty small rewordings and some reformatting. 

I think the level of exploration you have for the DESeq results is just about right. I would not want to go into much more depth there. 

I do wonder if we want to circle back to the fact that differential abundance can still be something you might want to look at, and maybe point to a chapter about it. I might says something about the kind of things you would want to think about before doing such an analysis, but not show any code about it.

I also started to anticipate some problems that might show up as part of #601; I think one thing is that we want to be congnizant of not making our variable names too long, just because the line length problem is likely to crop up, sometimes in ways that are hard to avoid.
Other times the solution will be as simple as converting from 
```
function(arg1,
         arg2,
         arg3)
```
to 
```
function(
  arg1,
  arg2,
  arg3
)
```

I tend to prefer the latter when the function name gets too long, or if there is an assignment of the results of the function, but I don't usually find it worth a comment.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -307,26 +306,37 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+However, we do not want to use corrected gene expression values for differential expression, we want to use the original raw counts that are present in the `counts` assay of the `SingleCellExperiment` object. 
+It is advised to only use the corrected values for any analyses being performed at the cell level, e.g. dimensionality reduction.
+In contrast, it is not advised to use corrected values for any analyses that are gene-based, such as differential expression or marker gene detection, because within-batch and between-batch gene expression differences are no longer preserved.
+See the [OSCA chapter on Using the corrected values](http://bioconductor.org/books/3.15/OSCA.multisample/using-corrected-values.html#using-corrected-values) for more insight.
```
I want to note that I've been linking version `3.16`. Should we all sync up here?

allyhawkins replied:
I'm going to change to `3.16` for the links present in this notebook.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -307,26 +306,37 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+See the [OSCA chapter on Using the corrected values](http://bioconductor.org/books/3.15/OSCA.multisample/using-corrected-values.html#using-corrected-values) for more insight.
+
+For the remainder of the DE analysis we will be using data found in the integrated object that has been uncorrected. 
+As a reminder, the uncorrected gene expression data, in the `counts` and `logcounts` assays, correspond to data that has been merged (the first step we walked through prior to integration) but has not been corrected.
```
The phrasing feels a but weird to say "the uncorrected data are the data that has not been corrected".. I'm not necessarily sure I've suggested better phrasing though!

```suggestion
As a reminder, the uncorrected gene expression data, found in the `counts` and `logcounts` assays, correspond to data that has been merged (the first step we walked through prior to integration) into the same SCE but not yet integrated.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -299,26 +299,37 @@ For this example we will focus on identifying DE genes in one cell type, the `Tu
--  Collapses gene expression counts by sample, so that samples, rather than cells, represent replicates.
+- Produces larger counts, which allows us to use standard normalization and differential expression methods used by bulk RNA-seq. 
+- Collapses gene expression counts by sample, so that samples, rather than cells, represent replicates.
 - Masks variance within a sample to emphasize variance across samples.
```
+1!

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -299,26 +299,37 @@ For this example we will focus on identifying DE genes in one cell type, the `Tu
-Our first step will be to pseudo-bulk our dataset to group our cells by cell type and by sample.
-This will create a new `SingleCellExperiment` (SCE) object that contains the pseudo-bulked counts for all genes. 
-We can subset this SCE to just contain our cell type of interest (tumor myoblasts) for input to differential expression analysis. 
+When we pseudo-bulk we want to produce a new counts matrix that has all genes and condenses all cells to include one group for each combination of cell type and sample.
```
Agreed.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -299,26 +299,37 @@ For this example we will focus on identifying DE genes in one cell type, the `Tu
-This will create a new `SingleCellExperiment` (SCE) object that contains the pseudo-bulked counts for all genes. 
-We can subset this SCE to just contain our cell type of interest (tumor myoblasts) for input to differential expression analysis. 
+When we pseudo-bulk we want to produce a new counts matrix that has all genes and condenses all cells to include one group for each combination of cell type and sample.
+Psuedo-bulking will create a new SCE object with new dimensions that contain only the pseudo-bulked counts for all genes. 
+We can then subset this SCE to just include our cell type of interest (tumor myoblasts) for input to the DE analysis. 
```
"with new dimensions" sort of sounds like we're adding more reduced dimensions, a thing we often do but are extremely not doing here. 
Again, not necessarily tied to this suggestion but I think it could it to be a bit more explicit in some way.

```suggestion
Psuedo-bulking will create a new SCE object that contains only the pseudo-bulked counts for all genes across all _samples_, rather than across all cells.
We can then subset this SCE to just include our cell type of interest (tumor myoblasts) for input to the DE analysis. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -328,125 +339,162 @@ counts_mtx
-# create new matrix, summing the counts for each gene across all cells in the specified group
+```{r pseudobulk matrix}
+# create pseudo-bulked matrix, summing the counts for each gene across all cells in the specified group
 pseudobulked_counts <- cbind(
   rowSums(counts_mtx[,c("A-cell1", "A-cell2")]), # sum all counts for cells labeled with group A
   rowSums(counts_mtx[,c("B-cell1", "B-cell2")]) # sum all counts for cells labeled with group B
 )
````
I support this idea, with an added comment that there is no base colsum (and also it's singular) so it's clear why we're using the function from this package.
That said, I have some memory that @jashapiro found some kind of bug or problem with this function? Did I make that up?

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -328,125 +339,162 @@ counts_mtx
 
-# column names aren't automatically added, so let's add them in 
+# column names aren't automatically added to the pseudo-bulked sce, so let's add them in 
 colnames(pseudobulked_sce) <- glue::glue("{pseudobulked_sce$celltype_broad}_{pseudobulked_sce$sample}")
```
I vote glue content next line for more glue!

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -328,125 +339,162 @@ counts_mtx
+### Perform differential expression with DESeq2
 
-Now we are ready to set up our `DESeq2` object. 
+Now we will use the `DESeq2` package to perform DE analysis on our pseudo-bulked SCE object.
```
+1

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -336,125 +346,162 @@ counts_mtx
+This is because we took the unnormalized counts from the original SCE to create the new pseudo-bulked counts which then become the input to create the `DESeq2` object.
+We don't want to sum the normalized counts because those have each been normalized within their sample and do not account for differences between samples.
+When we run `DESeq2` on our dataset, raw counts will first be normalized using size factors to account for differences in total sample counts.
+Therefore we don't have to do any normalization on our own and we can let `DESeq2` do all the work for us.
```
```suggestion
Therefore we don't have to do any normalization on our own - we'll let `DESeq2` do all the work for us.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -336,125 +346,162 @@ counts_mtx
-```{r}
-# estimate size factors first 
+However, we want to take a look at our data before we run the DE analysis.
+In particular, we can use principal component analysis (PCA) of our pseudo-bulked dataset as an initial test of our hypothesis that ARMS and ERMS samples have different gene expression patterns.
````
I'm a little unsure about calling this a test...maybe "exploration of our hypothesis"?


```suggestion
In particular, we can use principal component analysis (PCA) of our pseudo-bulked dataset as an initial exploration of our hypothesis that ARMS and ERMS samples have different gene expression patterns.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -336,125 +346,162 @@ counts_mtx
+If there is variation in their gene expression patterns we would expect samples from each RMS subtype to separate in PC space.
+We can evaluate this by plotting PC1 and PC2.
+
+In order to create our PCA plot, we will first need to normalize our data to account for any technical variations across samples.
+
+```{r normalize}
+# estimate size factors first
````
It feels odd to say that we don't have to normalize since DESeq takes care of it, but then we actually do normalize here. It seems like we always have to do size factors as well?
So, let's add more context about what functions we're running for DE specifically, and what we're running for exploratory viz.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
 
-Now we can get the updated results and combine with the `rowData` from the SCE object to extract the gene symbol. 
+If you look at our `shrink_results` object, we see that the genes are labeled with the ensembl gene identifiers, as those were the rownames of the pseudo-bulked SCE we used as input to build our `DESeq2` object.
+Although some of us may have all of the identifiers memorized by heart, it can be useful to have a human readable symbol in our results.
```
🎉

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
+
+Now that we have identified a set of genes that are DE in the tumor myoblasts between ARMS and ERMS subtypes, lets actually take a look at them and see if we can make some informative plots.
+The first plot we'll make is a volcano plot using the [`EnhancedVolcano` package](https://github.com/kevinblighe/EnhancedVolcano).
+This package automatically color codes the points by cutoffs for both significance and fold change and labels many of the significant genes (subject to spacing).
```
```suggestion
This package automatically colors the points by cutoffs for both significance and fold change and labels many of the significant genes (subject to spacing).
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -468,58 +515,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
+The first plot we'll make is a volcano plot using the [`EnhancedVolcano` package](https://github.com/kevinblighe/EnhancedVolcano).
+This package automatically color codes the points by cutoffs for both significance and fold change and labels many of the significant genes (subject to spacing).
+`EnhancedVolcano` has many, many options, which is a good thing if you don't like all of it's default settings.
+Even better, it outputs a `ggplot2` object, so if we want to customize it further, we can do that with the same `ggplot2` commands we have used before.
```
mixed antecedents here - 

```suggestion
Even better, it outputs a `ggplot2` object, so if we want to customize the plot further, we can use the same `ggplot2` commands we have used before.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
-
-We will pick a gene and look at the expression of that gene across subdiagnosis, sample, and cell type.
+We can also return back to the SCE object that we used to create our pseudo-bulked SCE and look at gene expression of some of the significant genes. 
+We can create UMAP plots as we did previously but instead of labeling each cell with metadata we can color each of the cells by the level of gene expression of a specified gene. 
```
```suggestion
We can create UMAP plots as we did previously, but instead of labeling each cell with metadata, we can color cells by a specified gene's expression levels.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+We will also use some of the `ggplot2` skills we picked up earlier, like `facet_grid()` to plot cells from different RMS subtypes separately.
+This can help us validate the `DESeq2` results so that we can visualize gene expression changes across our cell type of interest on a single-cell level. 
 
 ```{r}
````
chunk label

jashapiro replied:
All chunks below this need labels, it seems.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
 ```
 
+In the above plot we only plotted the tumor myoblast cells that we used in our DE analysis. 
+However, we might be interested to see the expression of genes that are differentially expressed in tumor myoblasts in other cell types present in our samples.
````
I think you mean...

```suggestion
However, we might be interested to see the expression of genes that are differentially expressed in other cell types present in our samples.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
 
 ```{r}
-# let's look across some other cell types
+# let's compare gene expression across some other cell types
````
any reason why these 4? Well, tumor is implicit, but why this specific normal tissue?

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -538,73 +595,66 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
 
+Next we will look at a few DE genes that we identified, one upregulated gene and one downregulated gene, and compare their expression in myoblasts to other cell types in ARMS and ERMS samples.
+We will use the `scater::plotExpression()` function to create a violin plot with RMS subtype on the x-axis and gene expression on the y-axis.
+We can continue using our `facet_grid()` functionality and show each cell type in its own column of the plot grid.
```
```suggestion
We can continue using `facet_grid()` to show separate panels for each cell type.
```

sjspielman reviewed (commented):
Next steps look good overall! Some comments for 2023 🚀

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -299,26 +299,37 @@ For this example we will focus on identifying DE genes in one cell type, the `Tu
+Throughout the notebook we have been working with an integrated dataset that contains corrected gene expression data and a corrected UMAP.
+However, we do not want to use corrected gene expression values for differential expression, we want to use the original raw counts that are present in the `counts` assay of the `SingleCellExperiment` object. 
+It is advised to only use the corrected values for any analyses being performed at the cell level, e.g. dimensionality reduction.
+In contrast, it is not advised to use corrected values for any analyses that are gene-based, such as differential expression or marker gene detection, because within-batch and between-batch gene expression differences are no longer preserved.
```
I think we should link this chapter here - http://bioconductor.org/books/3.16/OSCA.multisample/using-corrected-values.html
The Background section has a decent explanation.

allyhawkins commented:
I've incorporated many of the review suggestions and this is now ready for another round of reviews. 

The main changes I made were updating some of the text around using why we use raw counts, the pseudo-bulking, and the normalization. Nothing super major, just some slight reorganization to hopefully clarify and address some of the comments that were made. 

I also shortened variable names and used `pb` as shorthand for `pseudobulk` to hopefully avoid some spelling mistakes we will inevitably have. I also went through and tried to shorten any potentially long lines or comments. I think I got most of them in the notebook.

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+  ncol = 4,
+  dimnames = list(c("geneA", "geneB", "geneC"),
+                  c("A-cell1", "A-cell2", "B-cell1", "B-cell2"))
+  )
```
Minor formatting
```suggestion
)
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
-                                                  id = pseudobulked_groups)
+# create a new SCE object that contains 
+# the pseudo-bulked counts across the provided groups 
+pb_sce <- scuttle::aggregateAcrossCells(rms_sce, 
+                                                  id = pb_groups)
```
spacing
```suggestion
pb_sce <- scuttle::aggregateAcrossCells(rms_sce, 
                                        id = pb_groups)
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+# column names aren't automatically added to the pseudo-bulked sce, 
+# so let's add them in 
+colnames(pb_sce) <- glue::glue(
+  "{pb_sce$celltype_broad}_{pb_sce$sample}"
+  )
```
```suggestion
  "{pb_sce$celltype_broad}_{pb_sce$sample}"
)
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+### Perform differential expression with DESeq2
 
-Now we are ready to set up our `DESeq2` object. 
+Now we will use the `DESeq2` package to perform DE analysis on our pseudo-bulked SCE object.
```
Probably not needed, but we might as well spell it out again
```suggestion
Now we will use the `DESeq2` package to perform differential expression (DE) analysis on our pseudo-bulked SCE object.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+
+#### Create the `DESeq2` object
+
+To create the `DESeq2` object we will need the unnormalized counts matrix, the metadata associated with the samples, and a design formula. 
```
Adding a bit more detail about why we are using the sce object here.
```suggestion
To create the `DESeq2` object we will need the unnormalized counts matrix, the metadata associated with the samples, and a design formula.
The first two items are already stored in our SCE object, so we can create a DESeq2 object directly from that object using the `DESeqDataSet()` function.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+If you look at the pseudo-bulked SCE object you should have only seen one assay, the `counts` assay.
+This is because `DESeq2` expects raw counts.
+When we run `DESeq2` on our dataset, raw counts will first be normalized using size factors to account for differences in total sample counts.
+Therefore we don't have to do any normalization on our own - we'll let `DESeq2` do all the work for us.
```
Another pedantic change to an em-dash
```suggestion
Therefore we don't have to do any normalization on our own – we'll let `DESeq2` do all the work for us.
```

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+We expect to see the dispersion estimates decrease as means are increasing and follow the line of best fit. 
 
-```{r}
+```{r plot dipsersion}
````
```suggestion
```{r plot dispersion}
```

jashapiro reviewed (commented):
Looks good! I suggested mostly formatting changes, but also I wanted to make a different point about the effects of pseudobulking, which I tried to make more explicit in a comment, though I didn't attempt a rewrite. 

It looks like we need chunk labels for the end, but I realize there is probably one more PR here for the designation of live chunks, so that is not critical for this PR

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,228 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
-This will create a new `SingleCellExperiment` (SCE) object that contains the pseudo-bulked counts for all genes. 
-We can subset this SCE to just contain our cell type of interest (tumor myoblasts) for input to differential expression analysis. 
+This can be both good and bad! 
+Genes that are homogeneous across a sample are more likely to be identified as differentially expressed, while genes with high intra-sample heterogeneity will have a weaker signal.
```
Yes your example totally makes sense! I was thinking more along the lines if you have a gene that's uniformly up or down across cells then generally that would result in a stronger signal, whereas variation in your sample, even if just a small amount, will result in a weaker signal. Taking this from some of the explanation in the third bullet under the pseudo-bulking section of OSCA: http://bioconductor.org/books/3.16/OSCA.multisample/multi-sample-comparisons.html#creating-pseudo-bulk-samples

I'll update this to be more clear about what is meant by homogeneous.

allyhawkins replied:
I ended up removing the comment about homogeneous genes and reworded to emphasize that you would miss genes with intra-sample heterogeneous in https://github.com/AlexsLemonade/training-modules/pull/600/commits/ee2e73be65cf5a9fb1dd99f94bd9eed5fe773485 (and added in the missing chunk labels there too). Let me know what you think.

jashapiro replied:
I kind of partially disagree with part of their statement here, which is the point I was trying to get across: 

> Masking is generally desirable as DEGs - unlike marker genes - do not need to have low within-sample variance to be interesting, e.g., if the treatment effect is consistent across replicate populations but heterogeneous on a per-cell basis.

The part I disagree with is the implicit assumption that not knowing about heterogeneous cell expression is just fine. I kind of think it is worth knowing, and a model that ignores it seems weaker to me. But this is really subtle, and probably not worth the effort to try to get across in a short sentence. 

Maybe something like this? Or maybe just ignore it... This might be a thing for a slide where illustration of the point is easier.

```
Masking intra-sample variation means you might miss identifying genes where average expression doesn't change between samples but the degree of cell-to-cell variation does.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -309,154 +309,230 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
-```{r}
+As expected we see that samples group together based on RMS subtype and are separated along the PC1 axis, the PC contributing the highest amount of variation.
+
+We'll now use the wrapper function `DESeq()` to perform our differential expression analysis.
````
I think "wrapper" could be too jargon-y. We use "convenience function" for SingleR, does that work here?

```suggestion
We'll now use the convenience function `DESeq()` to perform our differential expression analysis.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -309,154 +309,230 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
 
-```{r}
+```{r plot dispersion}
 plotDispEsts(deseq_object)
````
```suggestion
plotDispEstsplotDispEsts(deseq_object)
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -470,58 +546,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
-# appyly logFC shrinkage
+```{r shrinkage}
+# appyly logFC shrinkage using the default model
 shrink_results <- DESeq2::lfcShrink(deseq_object, res = deseq_results, coef = 2, type = "apeglm")
````
This could go to multi-line as needed for spacing - 
 
```suggestion
shrink_results <- DESeq2::lfcShrink(deseq_object, 
                                                             res = deseq_results, 
                                                             coef = 2, 
                                                             type = "apeglm")
```

sjspielman replied:
Hm spacing isn't great over here in github, so you might want to let RStudio help!

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -470,58 +546,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
+```{r add gene symbol}
+# convert shrunken results to a tibble and add Ensembl id as a column
+deseq_results <- shrink_results |>
   tibble::as_tibble(rownames = "ensembl_id")
````
We might actually spend a second here - why not use `as.data.frame()`? b/c with this function we can convert the rownames to a column on the fly.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -470,58 +546,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
-The first plot we'll make is a volcano plot and we can label the significant genes with their gene symbols. 
+### Exploring the identified differentially expressed genes 
+
+Now that we have identified a set of genes that are DE in the tumor myoblasts between ARMS and ERMS subtypes, lets actually take a look at them and see if we can make some informative plots.
```
I don't really like the use of "DE" here since I feel like it stands for "differential expression" but here it's used as "differentially expressed." So, potentially, is this just me? If not..

```suggestion
Now that we have identified a set of genes that are differentially expressed in the tumor myoblasts between ARMS and ERMS subtypes, lets actually take a look at them and see if we can make some informative plots.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -470,58 +546,68 @@ We can correct this by applying a "shrinkage" procedure, which will adjust large
+### Exploring the identified differentially expressed genes 
+
+Now that we have identified a set of genes that are DE in the tumor myoblasts between ARMS and ERMS subtypes, lets actually take a look at them and see if we can make some informative plots.
+The first plot we'll make is a volcano plot using the [`EnhancedVolcano` package](https://github.com/kevinblighe/EnhancedVolcano).
```
I wonder if we should link bioc instead? If yes - 

```suggestion
The first plot we'll make is a volcano plot using the [`EnhancedVolcano` package](https://www.bioconductor.org/packages/3.16/bioc/html/EnhancedVolcano.html).
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -540,80 +626,82 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
+# create a violin plot 
 scater::plotExpression(tumor_sce,
-                       features = genes_to_plot, #PTPRT
+                       features = genes_to_plot, # a vector of genes to plot
```
This may run into spacing issues. If so, 
```suggestion
                       # a vector of genes to plot
                       features = genes_to_plot, 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -540,80 +626,82 @@ EnhancedVolcano::EnhancedVolcano(deseq_results,
                        point_size = 0.1) +
-  facet_grid(Feature ~ celltype_broad) + 
+  facet_grid(cols = vars(celltype_broad), # each celltype is its own column
+             rows = vars(Feature)) + # each feature is its own row
```
`scater::plotExpression()` actually creates this `Feature` column. We should explain this! Also this is why it doesn't have to be in the `other_fields` list.

sjspielman reviewed (commented):
This is coming along great! Let me know if any of my comments are unclear.

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,230 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+Now we are ready to start preparing for our DE analysis, where we will compare the gene expression of tumor myoblast cells between ARMS and ERMS samples.
+
+Throughout the notebook we have been working with an integrated dataset that contains corrected gene expression data and a corrected UMAP.
+However, we do not want to use corrected gene expression values for differential expression; we want to use the original raw counts that are present in the `counts` assay of the `SingleCellExperiment` object. 
```
We don't want `logcounts` for DESeq. It does not expect or use normalized data. We also aren't going to use logcounts for any kind of pseudobulking...

sjspielman replied:
This was not a clear comment from me and I meant to go back and change it, thanks for catching! I was trying to somehow communicate the difference between corrected and normalized since it felt odd to skip even talking about `logcounts` - including that we're not using it. Later I suggested to move a sentence up which is the better review!

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -309,154 +309,230 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
 ```
 
+Next we will create a pseudo-bulked matrix, with only 2 columns, 1 for group A and 1 for group B.
+To do this we will use the `DelayedArray::colsum()` function that allows us to sum the counts in all columns.
````
It's confusing, but this actually sums counts for each row, across columns. This may be a question of how my brain works. 

Maybe: 

```suggestion
To do this we will use the `DelayedArray::colsum()` function, which allows us to sum the counts for each row across groups of columns.
```

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -309,154 +309,230 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
-```{r}
-# note the new addition of the column with number of cells per group 
-colData(pseudobulked_sce)
+You should also see a new column that wasn't present previously, the `ncells` column.
+This column was added during pseudo-bulking and indicates the total number of cells that were summed together to form each column of the SCE object.
````
I moved it so that I first talk about the columns we've seen before, the QC columns and mention why they have NA. Then I talk about the new `ncells` column.

allyhawkins commented:
Okay I've incorporated most of the wording edits that were suggested, restructured the intro a little bit, added a small explanation in the multi-gene plot for why we can use `Feature`, and added in the sentence that Josh had suggested explaining the variation and heterogeneity issue. This should be ready for another look.

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -309,154 +309,232 @@ For this example we will focus on identifying DE genes in these `Tumor_Myoblast`
+As a reminder, the uncorrected gene expression data, found in the `counts` and `logcounts` assays, correspond to data that has been merged (the first step we walked through prior to integration) into the same SCE but not yet integrated.
+However, we do not want to use corrected gene expression values for differential expression; `DESeq2` expects the original raw counts as input so we will be using data found in the `counts` assay of the `SingleCellExperiment` object. 
+
+It is advised to only use the corrected values for any analyses being performed at the cell level, e.g. dimensionality reduction.
```
```suggestion
It is advised to only use the corrected values for any analyses being performed at the cell level, e.g., dimensionality reduction.
```

jashapiro reviewed (approved):
LGTM! (with one comma added)

sjspielman reviewed (approved):
🚀
